src/metabase/driver.clj

```diff
@@ -1294,6 +1304,31 @@
   dispatch-on-initialized-driver
   :hierarchy #'hierarchy)
 
+(defmulti insert-from-source!
+  "Inserts data from a data source into an existing table.
+   This abstracts away the conversion and insertion process, allowing drivers
+   to optimize based on the data source type. Returns number of rows inserted. "
```

Comment:
Extra docs / definition on 
- What are the requirements of implementers in terms of interpreting data sources (data source types they must implement)
- Is data being written expected to be visible to readers? (Or is a particular transaction + isolation level semantics being prescribed)
- If the operation fails is the implementation allowed to leave data in the table? (Same ACID analysis as above answers both I expect)
- Is there any limit to the size of the data source?
- Is there any requirement for implementations to timeout?
- Is the call expected to block and return only once finished with the data source (so callers must guarantee it exists only during the call)

Reply:
Totally agree - I would state it explicitly though 'Callers are free to X, are free to Y' etc